### PR TITLE
Remove all instances of low-level DH use in libssl (continued)

### PR DESCRIPTION
**Pull request by mattcaswell**
To prepare for the forthcoming deprecation of all the remaining low-level DH functions (see #13138), we remove all such usage from libssl.

There are a few instances where a DH object is passed as part of the public API. In those cases we deprecate the functions/macros where that is the case and provider EVP_PKEY based replacements.

(This is a continuation of PR #13233, which was closed due to github doing strange things)

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -23,6 +23,21 @@ OpenSSL 3.0
...
 
+ * The functions SSL_CTX_set_tmp_dh_callback and SSL_set_tmp_dh_callback, as
+   well as the macros SSL_CTX_set_tmp_dh() and SSL_set_tmp_dh() have been
+   deprecated. These are used to set the Diffie-Hellman (DH) parameters that
+   are to be used by servers requiring ephemeral DH keys. Instead applications
+   should consider using the built-in DH parameters that are available by
+   calling SSL_CTX_set_dh_auto() or SSL_set_dh_auto(). If custom parameters are
+   necessary then applications can use the alternative functions
+   SSL_CTX_set0_tmp_dh_pkey() and SSL_set0_tmp_dh_pkey(). There is no direct
+   replace for the "callback" functions. The callback was originally useful
```
Comment (slontis) [inline CHANGES.md:34]: replacement

File: crypto/bn/bn_dh.c [hunk 2]
```diff
@@ -10,7 +10,6 @@
 #include "bn_local.h"
 #include "internal/nelem.h"
 
-#ifndef OPENSSL_NO_DH
 # include <openssl/dh.h>
```
Comment (slontis) [inline crypto/bn/bn_dh.c:13]: Why are these constants not DH specific??

File: apps/s_server.c [hunk 3]
```diff
@@ -2030,54 +2022,63 @@ int s_server_main(int argc, char *argv[])
...
-        } else if (!SSL_CTX_set_tmp_dh(ctx, dh)) {
-            BIO_puts(bio_err, "Error setting temp DH parameters\n");
-            ERR_print_errors(bio_err);
-            DH_free(dh);
-            goto end;
+        } else {
+            if (!EVP_PKEY_up_ref(dhpkey)) {
+                EVP_PKEY_free(dhpkey);
+                goto end;
+            }
```
Comment (levitte) [inline apps/s_server.c:2046]: This looks *very* weird, and even more so the double `EVP_PKEY_free()`.  I assume that you simply want to "simulate" a `SSL_CTX_set1_tmp_dh_pkey()` call, but why?

What's worse, people "out there" use our apps source as examples of how to do things.  With that in mind, this seems over the top.  If you really have to do this, there should *at least* be a comment that explains this well.

File: crypto/bn/bn_dh.c (hunk 2 above)
Comment (levitte) [reply crypto/bn/bn_dh.c:13]: (me, I've always wondered why they are defined in the BN sub-library...)

**Comment by levitte** [comment]
Generally speaking, this is really good development, and long overdue.

**Comment by mattcaswell** [comment]
Updated to take account of the latest review comments. Also rebased to fix some conflicts with master.

File: crypto/bn/bn_dh.c (hunk 2 above)
Comment (levitte) [reply crypto/bn/bn_dh.c:13]: What I've always been thinking is that if BN was a separate, general purpose bignum project, we would *never* ask such a project to please insert *our* numbers into their code.  Now, that's not quite our reality, but if we're thinking in terms of layering and generality, have DH numbers in the BN sub-library is simply strange.

Comment (mattcaswell) [reply crypto/bn/bn_dh.c:13]: Well perhaps. But the fact is that at least some of these are exposed via BN_* calls. You can argue the rights and wrongs of this (and I think there are arguments to be made either way). But, no matter, I don't really view this as in scope of this PR (which is about removing low-level DH use in libssl) to refactor the BN<->DH layering (nor do I think that needs to be done for 3.0)

Comment (levitte) [reply crypto/bn/bn_dh.c:13]: Yes, sorry, agreed, it's off-topic for this PR

File: apps/s_server.c (hunk 3 above)
Comment (levitte) [reply apps/s_server.c:2050]: Thank you, that makes a world of a difference

File: util/other.syms [hunk 4]
```diff
@@ -476,6 +476,7 @@ SSL_CTX_set1_sigalgs_list               define
 SSL_CTX_set1_verify_cert_store          define
 SSL_CTX_set_current_cert                define
 SSL_CTX_set_ecdh_auto                   define
+SSL_CTX_set_dh_auto                     define
```
Comment (levitte) [inline util/other.syms:479]: Nit: are we trying to have some kind of alphabetical order in this file? :wink:

Comment (mattcaswell) [reply util/other.syms:479]: I'm redefining alphabetical order :-)

Comment (levitte) [reply util/other.syms:479]: Hmmm, so we should expect a new locale then?  :wink:
(en_GB.matt)

**Comment by mattcaswell** [comment]
Nit fixed. Please take another look.

**Review by levitte — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Comment by kaduk** [comment]
@nelsonopenssl IIUC this one should be "done", not just "ready to merge"?  See https://github.com/openssl/openssl/pull/13368#issuecomment-729720086

**Pull request closed without merging**